vnext/Desktop/WinHTTPWebSocketResource.cpp

```diff
@@ -0,0 +1,447 @@
+  );
+
+  if (!fStatus)
+  {
+    m_errorStatus = GetLastError();
+
+    if (m_errorHandler)
+    {
+      return m_errorHandler({ "Error upgrading protocol", ErrorType::Connection });
+    }
+  }
+
+  // Perform handshake
+  fStatus = WinHttpSendRequest(
+    m_requestHandle,                // hRequest
+    WINHTTP_NO_ADDITIONAL_HEADERS,  // lpszHeaders
+    0,                              // dwHeadersLength
+    NULL,                           // lpOptional
+    0,                              // dwOptionalLength
+    0,                              // dwTotalLength
+    (DWORD_PTR)this                 // dwContext
+  );
+  if (!fStatus)
+  {
+    m_errorStatus = GetLastError();
+
+    if (m_errorHandler)
+    {
+      return m_errorHandler({ "Error sending handshake", ErrorType::Handshake });
+    }
+  }
+  fStatus = WinHttpReceiveResponse(m_requestHandle, /*lpReserved*/ 0);
+  if (!fStatus)
+  {
+    m_errorStatus = GetLastError();
+
+    if (m_errorHandler)
+    {
+      return m_errorHandler({ "Error receiving handshake", ErrorType::Handshake });
+    }
+  }
+
+  // Complete upgrade. Accept only HTTP status code 101.
+  m_webSocketHandle = WinHttpWebSocketCompleteUpgrade(m_requestHandle, /*pContext*/ NULL);
+  if (m_webSocketHandle == NULL)
+  {
+    m_errorStatus = GetLastError();
+
+    if (m_errorHandler)
+    {
+      return m_errorHandler({ "Error completing protocol upgrade", ErrorType::Connection });
+    }
+  }
+
+  // Connection succeeded.
+  if (m_connectHandler)
+  {
+    m_connectHandler();
+  }
+}
+
+void WinHTTPWebSocketResource::Ping() {}
+
+void WinHTTPWebSocketResource::Send(const string& message)
+{
+}
+
+void WinHTTPWebSocketResource::SendBinary(const string& base64String)
+{
+}
+
+void WinHTTPWebSocketResource::Close(CloseCode code, const string& reason)
+{
+  auto result = WinHttpWebSocketShutdown(
+    m_webSocketHandle,                                         // hWebSocket
+    static_cast<USHORT>(code),                                 // usStatus
+    static_cast<LPWSTR>(_wcsdup(Utf8ToUtf16(reason).c_str())), // pvReason
+    static_cast<DWORD>(reason.length())                        // dwReasonLength
+  );
+  switch (result)
+  {
+  case ERROR_IO_PENDING:
+    GetLastError();
+    if (m_errorHandler)
+    {
+      m_errorHandler({ "Failed shutting donw WebSocket", ErrorType::Close });
+    }
+
+    break;
+  default:
+    // Success
+    GetLastError();//TODO:Remove
+    break;
+  }
+
+  result = WinHttpWebSocketClose(
+    m_webSocketHandle,                                        // hWebSocket
+    static_cast<USHORT>(code),                                // usStatus
+    static_cast<PVOID>(_wcsdup(Utf8ToUtf16(reason).c_str())), // pvReason
+    static_cast<DWORD>(reason.length())                       // dwReasonLength
+  );
+  switch (result)
+  {
+  case ERROR_INVALID_OPERATION:
+  case ERROR_INVALID_PARAMETER:
+    //case ERROR_INVALID_SERVER_RESPONSE: // 12152
+    //  GetLastError();
```

Comment:
do we not need this case? remove if so

Reply:
We hit it in runtime.
Strangely enough, online docs show it as `ERROR_INVALID_SERVER_RESPONSE`, but it doesn't seem to exist in the actual WinSDK headers.

Fixing...